### PR TITLE
darling: init at unstable-2023-05-02

### LINKED ISSUE #38628: Package darling
## Issue description
Darling is a compatibility layer for running macOS apps on Linux (think Wine for macOS). I think it would be interesting to work on with macOS support. They have reverse engineered some headers that Apple doesn't release publicly so having them in "libdarling" could be useful.

https://www.darlinghq.org/

### PR DESCRIPTION
**Pull request by zhaofengli**
###### Description of changes

This PR adds [Darling](https://github.com/darlinghq/darling) which is a Darwin/macOS emulation layer for Linux.

Patches applied:

- https://github.com/darlinghq/darling/pull/1355: Fixes the incorrect assumption during ELF parsing in the wrapper generator
- https://github.com/darlinghq/darling/pull/1356: Makes pre-compiling Python 2 bytecode optional, at the cost of performance
- https://github.com/darlinghq/darling/pull/1358: Removes libsystem_m.dylib's reference to the store
- https://github.com/darlinghq/darling-nano/pull/1: Fixes incorrect sysconfdir for nano
- https://github.com/darlinghq/darlingserver/pull/10: Allows overriding the path to xdg-user-dir

Fixes #38628.

## Installing

Set `programs.darling.enable = true;` in your NixOS config. This will install the setuid wrapper necessary to start Darling. Then, run `darling shell`.

## Stuff To Try

See also https://docs.darlinghq.org/what-to-try.html

```console
$ uname -a
Darwin host 19.6.0 Darwin Kernel Version 19.6.0 x86_64

# You can also use a real copy of Xcode (I haven't tried)
$ xcode-select --install
[...]

# Compile some code
$ clang hello.c

# Install Nix (does not fully work yet - more below)
$ curl -LO "https://github.com/zhaofengli/nix-on-darling/releases/download/initial-hack/nixos-system-x86_64-darwin.tar.xz"
[...]
$ tar xpf nixos-system-x86_64-darwin.tar.xz -C /
[...]
$ bootstrap-nix
Bootstrapping Nix...
installing 'nix-2.13.3'
building '/nix/store/ja3kqp1qwz66iqbkg50wzj264rppcnmg-user-environment.drv'...
installing 'nss-cacert-3.86'
building '/nix/store/27fw1kpgngj3q29ndphb30dw1vkhaxdz-user-environment.drv'...
Run the following command to make Nix work in the current shell:
. '/nix/var/nix/profiles/default/etc/profile.d/nix-daemon.sh'

$ . /nix/var/nix/profiles/default/etc/profile.d/nix-daemon.sh

$ nix-instantiate --eval -E 'builtins.currentSystem'
"x86_64-darwin"

# Try other stuff
$ nix run nixpkgs#neofetch
                    c.'          zhaofeng@host 
                 ,xNMM.          -------------- 
               .OMMMMo           OS: macOS Catalina 10.15 Darling x86_64   
               lMM"              Kernel: 19.6.0 
     .;loddo:.  .olloddol;.      Uptime: 86 days, 4 hours, 9 mins 
   cKMMMMMMMMMMNWMMMMMMMMMM0:    Shell: fish /nix/store/jawzvjj5x2iwv8zmz0905ix2v62a9h3r-neofetch- 
 .KMMMMMMMMMMMMMMMMMMMMMMMWd.    DE: Aqua 
 XMMMMMMMMMMMMMMMMMMMMMMMX.      WM: Quartz Compositor 
;MMMMMMMMMMMMMMMMMMMMMMMM:       WM Theme: Blue (Print: Entry, AppleInterfaceStyle, Does Not Exist 
:MMMMMMMMMMMMMMMMMMMMMMMM:       Terminal: shellspawn 
.MMMMMMMMMMMMMMMMMMMMMMMMX.      Memory: 27141MiB / 32034MiB 
 kMMMMMMMMMMMMMMMMMMMMMMMMWd.
 'XMMMMMMMMMMMMMMMMMMMMMMMMMMk                           
  'XMMMMMMMMMMMMMMMMMMMMMMMMK.                           
    kMMMMMMMMMMMMMMMMMMMMMMd
     ;KMMMMMMMWXXWMMMMMMMk.
       "cooc*"    "*coo'"
```

## Nix

The official Nix install script does not work on Darling due to its attempts to modify partitions and users on the system. Nix also has trouble [clearing file attributes](https://github.com/NixOS/nix/blob/7474a90db69813d051ab1bef35c7d0ab958d9ccd/src/libstore/local-store.cc#L607-L615) in the emulated filesystem.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/emulators/darling/default.nix [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    patchShebangs .
+
+    sed -i 's|SETUID||g' src/startup/CMakeLists.txt
+    sed -i 's|SETGID||g' src/external/basic_cmds/CMakeLists.txt
+  '';
+
+  nativeBuildInputs = [
+    ccWrapperBypass
+
+    python2
```
Comment (Atemu) [inline pkgs/applications/emulators/darling/default.nix:127]: :/

File: pkgs/applications/emulators/darling/default.nix [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+      hash = "sha256-gxXeNNutvEplpX8Wj2Sd17BxBWds+fmYo4vxs2Ip7vk=";
+    })
+  ];
+
+  postPatch = ''
+    chmod +x src/external/bootstrap_cmds/migcom.tproj/mig.sh
+    patchShebangs .
+
+    sed -i 's|SETUID||g' src/startup/CMakeLists.txt
+    sed -i 's|SETGID||g' src/external/basic_cmds/CMakeLists.txt
```
Comment (ShamrockLee) [inline pkgs/applications/emulators/darling/default.nix:121]:
```suggestion
    substituteInPlace src/startup/CMakeLists.txt --replace SETUID ""
    substituteInPlace src/external/basic_cmds/CMakeLists.txt --replace SETGID ""
```

`substituteInPlace` warns when the pattern is not found.

File: pkgs/applications/emulators/darling/default.nix (hunk 1 above)
Comment (Atemu) [reply pkgs/applications/emulators/darling/default.nix]: Neat, thanks for the option!

Comment (zhaofengli) [reply pkgs/applications/emulators/darling/default.nix]: I submitted [a patch](https://github.com/darlinghq/darling/pull/1356) upstream to make bytecode compilation optional. Disabling it does hurt performance, but it's not really used much so it's okay.

File: pkgs/applications/emulators/darling/default.nix [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+, libglvnd
+, fontconfig
+, dbus
+, libGLU
+, fuse
+, ffmpeg
+, pulseaudio
+
+, makeWrapper
+, python2 ? null
```
Comment (AndersonTorres) [inline pkgs/applications/emulators/darling/default.nix:24]:
```suggestion
, python2
```

File: pkgs/applications/emulators/darling/default.nix [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    python3
+
+    cmake
+    ninja
+
+    pkg-config
+
+    bison
+    flex
+  ] ++ lib.optional compilePy2Bytecode python2;
```
Comment (AndersonTorres) [inline pkgs/applications/emulators/darling/default.nix:149]:
```suggestion
  nativeBuildInputs = [
    bison
    ccWrapperBypass
    cmake
    flex
    makeWrapper
    ninja
    pkg-config
    python3
  ] 
  ++ lib.optional compilePy2Bytecode python2;
```

File: pkgs/applications/emulators/darling/default.nix [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+  cmakeFlags = [
+    "-DTARGET_i386=OFF"
+    "-DCOMPILE_PY2_BYTECODE=${if compilePy2Bytecode then "ON" else "OFF"}"
+  ];
+
+  NIX_CFLAGS_COMPILE = "-Wno-macro-redefined -Wno-unused-command-line-argument";
+
+  # Linux .so's are dlopen'd by wrapgen during the build
+  LD_LIBRARY_PATH = lib.makeLibraryPath wrappedLibs;
```
Comment (AndersonTorres) [inline pkgs/applications/emulators/darling/default.nix:172]:
```suggestion
  env.NIX_CFLAGS_COMPILE = "-Wno-macro-redefined -Wno-unused-command-line-argument";

  # Linux .so's are dlopen'd by wrapgen during the build
  env.LD_LIBRARY_PATH = lib.makeLibraryPath wrappedLibs;
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by zhaofengli** [comment]
The previous `patchShebangs .` was too indiscriminate and caused scripts intended to be run inside the Darling sandbox to be patched as well. Also added a check to make sure that the Darling sysroot doesn't contain references to the Nix store (except for mldr).

File: pkgs/applications/emulators/darling/default.nix [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+    freetype
+    libjpeg
+    libpng
+    libtiff
+    giflib
+    xorg.libX11
+    xorg.libXext
+    xorg.libXrandr
+    xorg.libXcursor
+    xorg.libxkbfile
```
Comment (AndersonTorres) [inline pkgs/applications/emulators/darling/default.nix:88]:
```suggestion
    libX11
    libXext
    libXrandr
    libXcursor
    libxkbfile
```

xorg was deprecated a short time ago.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: nixos/modules/programs/darling.nix [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  options = {
+    programs.darling = {
+      enable = mkOption {
+        default = false;
+        type = types.bool;
+        description = lib.mdDoc ''
+          Whether to set up Darling, a Darwin/macOS compatibility layer
+          for Linux.
+        '';
+      };
```
Comment (Kranzes) [inline nixos/modules/programs/darling.nix:15]:
```suggestion
      enable = mkEnableOption (lib.mdDoc "Darling, a Darwin/macOS compatibility layer for Linux");
```

File: nixos/modules/programs/darling.nix [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    };
+  };
+
+  config = mkIf config.programs.darling.enable {
+    security.wrappers.darling = {
+      source = "${pkgs.darling}/bin/darling";
```
Comment (Kranzes) [inline nixos/modules/programs/darling.nix:21]:
```suggestion
      package = mkPackageOptionMD pkgs "darling" {};
    };
  };

  config = mkIf config.programs.darling.enable {
    security.wrappers.darling = {
      source = lib.getExe config.programs.darling.package;
```

**Comment by zhaofengli** [comment]
During the build, Darling produces a source-built SDK which is now installed into a separate output. This SDK can be used to cross-compile from Linux to macOS (see the NixOS test). I tried the resulting executable and it worked fine on a real M1 Mac mini under Rosetta 2.

<details>
<summary>Very hacky cross stdenv</summary>

```nix
let
  lib = import ./lib;

  pkgs = import ./. {
    crossSystem = lib.systems.elaborate (lib.systems.examples.x86_64-darwin // {
      darwinSdkVersion = "10.15";
      darwinMinVersion = "10.15";
    });

    overlays = [ (final: prev: {
      stdenv =
        if (prev.stdenv.hostPlatform == prev.stdenv.buildPlatform) then prev.stdenv
        else prev.stdenv.override {
          overrides = overrideStdenv;
        };
    }) ];
  };

  overrideStdenv = final: prev: let
    nativePkgs = prev.pkgsBuildBuild;

    llvm = nativePkgs.llvmPackages_latest;
    sdk = "${nativePkgs.darling.sdk}/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk";

    nativeStuff = {
      nativeTools = false;
      nativeLibc = false;
      nativePrefix = "";
      shell = nativePkgs.stdenvNoCC.shell;
      inherit (nativePkgs) coreutils gnugrep;
    };

    mergedBintools = nativePkgs.symlinkJoin {
      name = "bintools-unwrapped";
      paths = [ llvm.bintools-unwrapped ];
      postBuild = ''
        ln -sf ${nativePkgs.darling.sdk}/bin/x86_64-apple-darwin19-ld $out/bin/ld
      '';
    };

    bintools = prev.wrapBintoolsWith ({
      bintools = mergedBintools;
      libc = "${sdk}/usr";
    } // nativeStuff);

    cc = prev.wrapCCWith (rec {
      cc = llvm.clang-unwrapped;
      inherit bintools;
      libc = "${sdk}/usr";
      libcxx = null;

      isClang = true;

      gccForLibs = null;
      extraBuildCommands = ''
        echo "--sysroot=${sdk}" >> $out/nix-support/cc-cflags

        rsrc="$out/resource-root"
        mkdir "$rsrc"
        ln -s "${cc.lib or cc}/lib/clang/${cc.version}/include" "$rsrc"
        echo "-resource-dir=$rsrc" >> $out/nix-support/cc-cflags
      '';
    } // nativeStuff);

    stdenv = prev.stdenv.override {
      inherit cc;
      extraBuildInputs = [];
    };
  in {
    inherit stdenv;
  };
in pkgs
```
</details>

**Comment by zhaofengli** [comment]
@ofborg test darling

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/applications/emulators/darling/default.nix [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+    cairo
+    libglvnd
+    fontconfig
+    dbus
+    libGLU
+
+    # darling-dmg
+    fuse
+
+    # CoreAudio
```
[the remark this answers is not in the document]
Comment (zhaofengli) [reply pkgs/applications/emulators/darling/default.nix:102]: If you are asking about the comments, these are supposed to indicate which components the wrapped libraries are used in. I'll make that clearer.

**Pull request merged**